sweep/fee_bumper.go

```diff
@@ -1277,28 +1443,93 @@ func (t *TxPublisher) createAndPublishTx(
 // a tx not known to us. When a tx is not confirmed, yet its inputs has been
 // spent, then it must be spent by a different tx other than the sweeping tx
 // here.
+//
+// NOTE: This method will update `r.tx` if a confirmed wallet tx is found.
 func (t *TxPublisher) isUnknownSpent(r *monitorRecord,
 	spends map[wire.OutPoint]*wire.MsgTx) bool {
 
-	txid := r.tx.TxHash()
+	// walletTx is the confirmed tx found in our wallet. When it's set, it
+	// means  a previous tx has spent the inputs, which could happen during
+	// restarts.
+	var walletTx *wire.MsgTx
 
 	// Iterate all the spending txns and check if they match the sweeping
 	// tx.
 	for op, spendingTx := range spends {
-		spendingTxID := spendingTx.TxHash()
+		spendingTxid := spendingTx.TxHash()
 
 		// If the spending tx is the same as the sweeping tx then we are
 		// good.
-		if spendingTxID == txid {
+		if r.tx != nil && spendingTxid == r.tx.TxHash() {
 			continue
 		}
 
+		// Fetch the wallet to see if it's our previous sweeping tx.
+		tx, err := t.cfg.Wallet.FetchTx(spendingTxid)
+		if err != nil {
+			log.Errorf("Failed to FetchTx for %v: %v", spendingTxid,
+				err)
+
+			continue
+		}
+
+		// Found a wallet tx, which means a previous sweeping tx is now
+		// confirmed.
+		if tx != nil {
+			// Assign the tx for the first time.
+			if walletTx == nil {
+				walletTx = tx
+				continue
+			}
+
+			// Inputs are spent via the same tx, which is the
+			// expected case.
+			if walletTx.TxHash() == tx.TxHash() {
+				continue
+			}
+
+			// We'd end up here if the inputs are spent by at least
+			// two previous sweeping txns, which could happen if the
+			// two sweeping txns are spending non-time sensitive
+			// inputs:
+			// 1. sweeping tx 1 published and stays in the mempool.
+			// 2. sweeping tx 2 published and stays in the mempool.
+			// 3. the node shuts down.
+			// 4. the two sweeping txns confirmed.
+			// 5. the node comes back and reoffers the inputs, which
+			//    could be grouped together if they all have non
+			//    deadline heights.
+			//
+			// This is fine as we will return the first tx and retry
+			// the sweeping process, which will end up here again
+			// with a deducted input set.
```

Comment:
Comment seems incorrect.  IIUC we will fall through to the below case and return true, which means all the spent inputs will be marked as "unknown spent".  Then the sweeper will remove them from the next batch, and probably we *won't* hit this case again.

Reply:
Upon receiving unknown spends the sweeper will handle them by checking the spending tx against the sweeper store - if it's there, they will be marked as swept, otherwise fatal - since the inputs are already spent, they cannot be retried again.